iceoryx_utils/include/iceoryx_utils/concurrent/resizeable_lockfree_queue.hpp

```diff
@@ -0,0 +1,198 @@
+// limitations under the License.
+
+#ifndef IOX_UTILS_CONCURRENT_RESIZEABLE_LOCKFREE_QUEUE_HPP
+#define IOX_UTILS_CONCURRENT_RESIZEABLE_LOCKFREE_QUEUE_HPP
+
+#include "iceoryx_utils/concurrent/lockfree_queue.hpp"
+#include "iceoryx_utils/cxx/type_traits.hpp"
+#include "iceoryx_utils/cxx/vector.hpp"
+
+#include <atomic>
+
+namespace iox
+{
+namespace concurrent
+{
+/// @brief implements a lock free queue (i.e. container with FIFO order) of elements of type T
+/// with a maximum capacity MaxCapacity.
+/// The capacity can be defined to be anything between 0 and MaxCapacity at construction time
+/// or later at runtime using setCapacity. This is even possible while concurrent push and pop
+/// operations are executed, i.e. the queue does not have to be empty.
+/// Only one thread will succeed setting its desired capacity if there are more
+/// threads trying to change the capacity at the same time (it is unpredictable which thread).
+
+// Remark: We use protected inheritance to make the base class methods inaccessible for the user.
+// We cannot use virtual functions since we need to use this class in shared memory.
+// Some of the methods need to be rewritten specifically for this class, others simply redirect
+// the call to the base class.
+//
+// Since supporting the resize (setCapacity) functionality has an impact on the runtime even
+// if the feature is not used, we provide a queue wihout resize functionality in an additional
+// base class that can be used separately.
+template <typename ElementType, uint64_t MaxCapacity>
+class ResizeableLockFreeQueue : protected LockFreeQueue<ElementType, MaxCapacity>
+{
+  private:
+    using Base = LockFreeQueue<ElementType, MaxCapacity>;
+
+  public:
+    using element_t = ElementType;
+    static constexpr uint64_t MAX_CAPACITY = MaxCapacity;
+
+    ResizeableLockFreeQueue() = default;
+    ~ResizeableLockFreeQueue() = default;
+
+    // deleted for now, can be implemented later if needed
+    // note: concurrent copying or moving in lockfree fashion is nontrivial
+    ResizeableLockFreeQueue(const ResizeableLockFreeQueue&) = delete;
+    ResizeableLockFreeQueue(ResizeableLockFreeQueue&&) = delete;
+    ResizeableLockFreeQueue& operator=(const ResizeableLockFreeQueue&) = delete;
+    ResizeableLockFreeQueue& operator=(ResizeableLockFreeQueue&&) = delete;
+
+    ResizeableLockFreeQueue(uint64_t initialCapacity) noexcept;
+
+    /// @brief returns the maximum capacity of the queue
+    /// @return the maximum capacity
+    static constexpr uint64_t maxCapacity() noexcept;
+
+    /// @brief tries to insert value in FIFO order, copies the value internally
+    /// @param[in] value to be inserted
+    /// @return true if insertion was successful (i.e. queue was not full during push), false otherwise
+    /// @note threadsafe, lockfree
+    /// bool tryPush(ElementType&& value) noexcept;
+    /// bool tryPush(const ElementType& value) noexcept;
+    using Base::tryPush;
```

Comment:
Yes, I was considering this but I thought it may be annoying having to look it up in the base class. Ok, will remove. The whole design is a bit unconventional (the way of reimplementing some methods, but keeping other ones as in base but without virtual), for good reason though.

I can remove it, it is not important.